docs/language-reference/04-types.md

```diff
@@ -0,0 +1,323 @@
+
+* If a field would have _improper straddle_, where the interval `(fieldOffset, fieldOffset+fieldSize)` (eclusive on both sides) contains any multiple of 16, *and* the field offset is not already a multiple of 16, then the offset of the field is adjusted to the next multiple of 16
+
+Array Types
+-----------
+
+An _array type_ is either a statically-sized or dynamically-sized array type.
+
+A known-size array type is written `T[N]` where `T` is a type and `N` is a specialization-time constant integer.
+This type represents an array of exactly `N` values of type `T`.
+
+An unknown-size array type is written `T[]` where `T` is a type.
+This type represents an array of some fixed, but statically unknown, size.
+
+### Declaration Syntax
+
+For variable and parameter declarations using traditional syntax, a variable of array type may be declared by using the element type `T` as a type specifier (before the variable name) and the `[N]` to specify the element count after the variable name:
+
+```hlsl
+int a[10];
+```
+
+Alternatively, the array type itself may be used as the type specifier:
+
+```hlsl
+int[10] a;
+```
+
+When using the `var` or `let` keyword to declare a variable, the array type must not be split:
+
+```hlsl
+var a : int[10];
+```
+
+> Note: when declaring arrays of arrays (often thought of as "multidimensional arrays") a programmer must be careful about the difference between the two declaration syntaxes.
+> The following two declarations are equivalent:
+>
+> ```hlsl
+> int[3][5] a;
+> int a[5][3];
+> ```
+>
+> In each case, `a` is a five-element array of three-element arrays of `int`s.
+> However, one declaration orders the element counts as `[3][5]` and the other as `[5][3]`.
+
+### Element Count Inference
+
+When a variable is declared with an unknown-size array type, and also includes an initial-value expression:
+
+```hlsl
+int a[] = { 0xA, 0XB, 0xC, 0xD };
+```
+
+The compiler will attempt to infer an element count based on the type and/or structure of the initial-value expression.
+In the above case, the compiler will infer an element count of 4 from the structure of the initializer-list expression.
+Thus the preceding declaration is equivalent to:
+
+```hlsl
+int a[4] = { 0xA, 0xB, 0xC, 0xD };
+```
+
+A variable declared in this fashion semantically has a known-size array type and not an unknown-size array type; the use of an unknown-size array type for the declaration is just a convenience feature.
+
+### Standard Layout
+
+The _stride_ of a type is the smallest multiple of its alignment not less than its size.
+
+Using the standard layout for an array type `T[]` or `T[N]`:
+
+* The _element stride_ of the array type is the stride of its element type `T`
+* Element `i` of the array starts at an offset that is `i` times the element stride of the array
+* The alignment of the array type is the alignment of `T`
+* The size of an unknown-size array type is unknown
+* The size of a known-size array with zero elements is zero
```

Comment:
Yeah, our reflection and layout stuff might kludge things and treat a zero-element array as being an unknown-size array. We should fix that at some point.

(And yes, the meta-level point is that at times this document veers into being aspirational rather than just documenting the status quo. We should probably do better about calling out things that ought to work but don't).

Zero-element arrays should be easy enough to support in principle. We just need an IR pass that removes all variables/fields/parameters with zero-element-array type and any attempts to index into a zero-element array with `undefined` values.